go-controller/pkg/ovn/base_network_controller.go

```diff
@@ -532,24 +532,30 @@ func (bnc *BaseNetworkController) deleteNamespaceLocked(ns string) (*namespaceIn
 		}
 
 		// Delete the address set after a short delay.
-		// This is so NetworkPolicy handlers can converge and stop referencing it.
+		// This is to avoid OVN warnings while the address set is still
+		// referenced from NBDB ACLs until the NetworkPolicy handlers remove
+		// them.
 		addressSet := nsInfo.addressSet
 		go func() {
 			select {
 			case <-bnc.stopChan:
 				return
 			case <-time.After(20 * time.Second):
-				// Check to see if the NS was re-added in the meanwhile. If so,
-				// only delete if the new NS's AddressSet shouldn't exist.
-				nsInfo, nsUnlock := bnc.getNamespaceLocked(ns, true)
-				if nsInfo != nil {
-					defer nsUnlock()
-					if nsInfo.addressSet != nil {
+				bnc.namespacesMutex.Lock()
+				nsInfo := bnc.namespaces[ns]
+				if nsInfo == nil {
+					defer bnc.namespacesMutex.Unlock()
+				} else {
+					bnc.namespacesMutex.Unlock()
+					nsInfo.Lock()
+					defer nsInfo.Unlock()
+					bnc.namespacesMutex.Lock()
+					defer bnc.namespacesMutex.Unlock()
+					if nsInfo != bnc.namespaces[ns] || nsInfo.addressSet != nil {
```

Comment:
if we consider that a namespace may exist without an address set, then `nsInfo != bnc.namespace[ns]` could mean that namespace was re-created twice, and both nsInfos didn't have an address set, then we still need to clean up instead of return.
Generic logic for `if nsInfo != bnc.namespaces[ns] {return}` doesn't work here because this is a cleanup and not something that will be retried on error like add/update/delete namespace.